File: frontend/src/types/report_block.ts
/**
 * 分析报告的「块流」契约（ReportBlock）：文/图穿插叙事（2026-09-15 定稿）。
 *
 * 报告正文 = 有序的 text/chart 块序列，前端按序渲染、图穿插在段落之间。与后端
 * `app/schemas/report_block.py` **字段一一对应**（双侧同构），改一边必须改另一边。
 * 坏 chart 块单独丢弃、text 块照显（兜底红线）。
 */

import { isChartSpec, type ChartSpec } from '@/types/chart'

export interface TextBlock {
  kind: 'text'
  md: string
}

export interface ChartBlock {
  kind: 'chart'
  spec: ChartSpec
}

export type ReportBlock = TextBlock | ChartBlock

/**
 * 运行时校验 + 清洗：把后端/模型给的原始块数组洗成可安全渲染的 ReportBlock[]。
 * - 非对象 / kind 不认得 → 丢。
 * - text 块：md 非字符串 → 丢；空串也丢（没内容别占位）。
 * - chart 块：spec 过 isChartSpec 才留（模型配错图 → 只少这张图，不崩）。
 */
export function toReportBlocks(raw: unknown): ReportBlock[] {
  if (!Array.isArray(raw)) return []
  const out: ReportBlock[] = []
  for (const b of raw) {
    if (typeof b !== 'object' || b === null) continue
    const kind = (b as { kind?: unknown }).kind
    if (kind === 'text') {
      const md = (b as { md?: unknown }).md
      if (typeof md === 'string' && md.trim() !== '') out.push({ kind: 'text', md })
    } else if (kind === 'chart') {
      const spec = (b as { spec?: unknown }).spec
      if (isChartSpec(spec)) out.push({ kind: 'chart', spec })
    }
  }
  return out
}

File: app/schemas/report_block.py
"""分析报告的「块流」契约（ReportBlock）：文/图穿插叙事的稳定形状（2026-09-15 定稿）。

报告正文不再是「一个 body 字符串 + 尾部一摞 charts」，而是**有序的块序列**——analyze Agent
按叙述顺序产出 `text` / `chart` 块，前端按序渲染，图自然穿插在段落之间（讲一段→配一张→
再讲一段）。这比「正文 + 图集」更贴合「分维度逐段讲」的分析形态。

兜底红线不变：坏 `chart` 块（isChartSpec 不过）**单独丢弃、text 块照显**——模型配错图只少
一张图，绝不拖垮整段文字。双侧同构：本文件（Python）⇄ `frontend/src/types/report_block.ts`（TS）。
"""

from typing import Annotated, Literal

from pydantic import BaseModel, Field

from app.schemas.chart import ChartSpec


class TextBlock(BaseModel):
    """一段 Markdown 文字。"""

    kind: Literal["text"] = "text"
    md: str = ""


class ChartBlock(BaseModel):
    """一张图（内嵌 ChartSpec，走与独立配图同一份契约/校验）。"""

    kind: Literal["chart"] = "chart"
    spec: ChartSpec


# 判别联合：按 kind 分流。报告正文 = 这个块的有序列表（0~N 块，宁缺勿滥）。
ReportBlock = Annotated[TextBlock | ChartBlock, Field(discriminator="kind")]
